README.md
# Project Moonshot

Feel free to read the code.

## Overview of Contracts 

### Moonshot.sol BEP20 token

Deflationary, Frictionless yield and liquidity generation protocol. Simply hold Moonshot in your wallet and gain rewards when others transfer.

The contract takes a 10% fee:
 - 4% is redistributed to all holders
 - 3% is added to the V2 LP pool
 - 3% is added to the Moonshot support fund

Notes:
- The contract contains functions for the owner to blacklist addresses 
- BNB sent to the contract can be rescued
- There is no max TX amount and the total fees can be set to at most 10%
- The router can be upgraded

### ClaimMoonshot.sol

Contract that sends MoonshotV2 to the caller if the caller has a positive Moonshot balance. The caller can only claim once.

Notes:
- The contract contains functions for the owner to black list addresses
- The contract contains a function for the owner to burn the token's MoonshotV2 balance 
- The contract contains a function to rescue BNB
- The contract contains no functions to withdraw the balance of a BEP20 token

### BuyMoonshot.sol

Contract that sends MoonshotV2 to the caller upon receiving BNB.
A 0.5% fee is deducted from the BNB received before doing a market buy of the Moonshot V2 token using the pancakeswap V2 Router

Notes:
- The contract contains a function to enable or disable the fee
- The contract contains a function to set the fee to at most 2%
- The contract contains functions to update token contract addresses
- The contract contains a function to withdraw BNB
- The contract contains a function to withdraw BEP20 tokens

## Setting up your workspace

1. run 'npm install' in your workspace
2. run Ganache and setup a workspace
3. run 'npm test' in your workspace

### Deploying locally

For testing:
1. run 'npx ganache-cli --deterministic --allowUnlimitedContractSize --networkId 2'
2. run 'npx truffle migrate --network test'

For development:
1. run 'npx ganache-cli --deterministic --allowUnlimitedContractSize --networkId 1 -p 7545'
2. run 'npx truffle migrate --network develop'

## Deploying on Smart Chain Testnet 

0. Create a wallet and get some BNB from the faucet: https://testnet.binance.org/faucet-smart
1. Copy secrets.env.template to secrets.env and set values
2. Run 'npx truffle compile'
3. Run 'npx truffle migrate --network bsctestnet'
4. Smart contract should now be deployed.
5. Verify the contract
6. Run 'npx truffle run verify moonshot@0x0000TOKENADDRES0000 --network bsctestnet

## Running scripts

To configure the contract instances for use, you can run the files in scripts/ 

0. Run 'npx truffle exec scripts/pre-finalize.js --network bsctestnet'



## License

WTFPL


RELEASE_NOTES.md
# Moonshot v2 0x5298ad82dd7c83eeaa31dda9deb4307664c60534

The moonshot token is migrating to V2

Old address: 0xd27d3f7f329d93d897612e413f207a4dbe8bf799
New address: 0x5298ad82dd7c83eeaa31dda9deb4307664c60534

## Why Migration

1. The upgrade to V2 allows us more autonomy over the liquidity pool and takes us away from V1 completely
2. We can offer enhanced security and remove the BitMart Hacker from the holder's list
3. We can resolve issues in the existing smart contract

## How to migrate your Moonshot

You can claim your Moonshot V1 balance as Moonshot V2 here: https://project-moonshot.me/moonswap

## Benefits of the Migration

Below follow a list of changes compared to the old Moonshot contract deployed at 0xd27d3f7f329d93d897612e413f207a4dbe8bf799

### Token Symbol "MSHOT"

The token symbol was changed from MOONSHOT to MSHOT 

### Enhanced Security

Moonshot can be more "reactive" to threats because of the ability to blacklist addresses.

### Liquidity

Adding to the Liquidity Pool generates LP tokens. These tokens accumulate over time in the owner's wallet (Deployer) and pose a risk to the project.
The Moonshot V2 contract changes that; The LP tokens will accumulate in the contract instead and a removeLiquidity function was added that allows
the owner to remove at most 10% of the liquidity every 4 weeks.

### Buy Back mechanism

The token contract has an optional buy back mechanism. When the contract buys Moonshot V2, the tokens are sent to the NULL address.

### Fee Mechanism

The total fees are the same as the old contract, but now the 6% reserved for adding to the Liquidity Pool was reduced to 3% and the remaining 3% is
sent directly to the team wallet to support the project's development / marketing costs . 

Old Fee structure: 4/6  (4% reflection, 6% LP )
New Fee stucture: 4/3/3 (4% reflection, 3% LP , 3% dev/marketing )

Fees cannot be set higher than 10% in total

Fees can be set with 2 decimal precision.

### Upgradable

The addresses for the Pancake Router and Pancake pair can be upgraded

### Set Max TX Amount

This function was removed

### Rescue BNB

BNB sent to the contract by mistake may be rescued. 


### Smaller changes

- Each function that changes the state of the contract emits an event to be recorded on chain
- Declared functions external where possible to reduce gas cost
- Resolved "Temporary Ownership Renounce" vulnerability 
- Resolved frictionless yield bug
- Removed not needed function 'deliver'
- Removed unused contracts
- Declared some variables as constant
- Removed redundant code
- Fix excluded[] length problem
- Fixed incorrect error messages 
- Naming things, fixed typos



truffle-config.js
require('dotenv').config();
const HDWalletProvider = require("truffle-hdwallet-provider");


module.exports = {
  mocha: {
    enableTimeouts: false,
  },
  networks: {
    develop: {
      host: "127.0.0.1",
      port: 8545,
      network_id: "1",
      gasPrice: 0,
      gas: 6000000,
    },
    test: {
      host: "127.0.0.1",
      port: 7545,
      network_id: "2"
    },
    bsctestnet: {
        provider: () => new HDWalletProvider( process.env.PRIVKEY, 'https://data-seed-prebsc-1-s1.binance.org:8545' ),
            network_id: 97,
            confirmations: 2,
            timeoutBlocks: 200,
            gas: 1000000,
            skipDryRun: true
    },
    ropsten: {
        provider: () => new HDWalletProvider( process.env.PRIVKEY, process.env.ROPSTEN_URL ),
            network_id: 3,
            confirmation: 10,
            timeoutBlocks: 200,
          //  gas: 10000000,
            skipDryRun: true,
    },
    bsc: {
        provider: () => new HDWalletProvider( process.env.PRIVKEY, 'https://bsc-dataseed1.binance.org'),
            network_id: 56,
            confirmations: 2,
            timeoutBlocks: 200,
          //  gas: 20000000,
            skipDryRun: true
    }
  },
  plugins: [
      'truffle-plugin-verify',
      'truffle-contract-size'
  ],
  api_keys: {
      bscscan: process.env.APIKEY,
      etherscan: process.env.ETHERSCAN_APIKEY
  },

  compilers: {
    solc: {
      version: "0.6.12",
      settings: {
        optimizer: {
          enabled: true,
          runs: 1000
        }
      }
    }
  }
};


scripts/pre-finalize.js
const Moonshot = artifacts.require( "Moonshot" );
const ClaimMoonshot = artifacts.require( "ClaimMoonshot" );
const BuyMoonshot = artifacts.require( "BuyMoonshot" );

module.exports = async function(callback) {
  try {
        
      const m = await Moonshot.deployed();
      const a = await ClaimMoonshot.deployed();
      const t = await BuyMoonshot.deployed();

      console.log("Moonshot at ", m.address);
      console.log("ClaimMoonshot at ", a.address);
      console.log("Buy contract at ", t.address);

      // set moonshot fund address and exclude from fee
      await m.setMoonshotFundAddress("0x9d8a5d6B405c2Eb7cee724F4B2F67a902F0f0864");
      await m.setFee("0x9d8a5d6B405c2Eb7cee724F4B2F67a902F0f0864", 0, true);
 
      // bitmart hotwallet
      await m.setFeeForAddress( "0x328130164d0F2B9D7a52edC73b3632e713ff0ec6", 0);
      await m.excludeFromReward( "0x328130164d0F2B9D7a52edC73b3632e713ff0ec6" );

      // exclude hotbit wallet
      await m.setFeeForAddress( "0xc7029e939075f48fa2d5953381660c7d01570171", 0);
      await m.excludeFromReward( "0xc7029e939075f48fa2d5953381660c7d01570171");

      // latoken
      await m.setFeeForAddress( "0xCE55977E7B33E4e5534Bd370eE31504Fc7Ac9ADc",0 );
      await m.excludeFromReward( "0xCE55977E7B33E4e5534Bd370eE31504Fc7Ac9ADc" );

      // p2pb2b
      await m.setFeeForAddress( "0x5be909E0D204A94cc93fc9D7940584B5EC59e618",0 );
      await m.excludeFromReward( "0x5be909E0D204A94cc93fc9D7940584B5EC59e618" );

      // moonshot
      await m.setFeeForAddress( "0x66A7B9f608378e59105022aB00b0F541666e8c4d", 0);
      await m.excludeFromReward( "0xe539958C08477B8EaBE2290e8E06dc657102572D");
 
      // whitelist moonshot claim contract
      await m.excludeFromReward( a.address );
      // users that claim are not subject to fee
      await m.setFee( a.address, 0, true );

      // set to token contract address in claim contract
      await a.setToTokenAddress( m.address );
      // set from token contract address in claim contract
      await a.setFromTokenAddress( "0xd27d3f7f329d93d897612e413f207a4dbe8bf799" );
     
      // add bitmart hacker to blacklist of claim contract
      await a.addToBlackList( "0x25fb126B6c6B5c8EF732b86822fA0F0024E16C61" );
      
      

      // set token address in buy contract
      await t.setTokenAddress( m.address );
      // set fee enabled in buy contract (0.5%)
      await t.setFeeEnabled( true );



      console.log("Add Liquidity, Burn and Fund Claim contract");
  }
  catch( error ) {
      console.log(error);
  }

  callback();
}
